source/darwin/dispatch_queue_event_loop.c

```diff
@@ -0,0 +1,686 @@
+    struct aws_uuid uuid;
+    AWS_FATAL_ASSERT(aws_uuid_init(&uuid) == AWS_OP_SUCCESS);
+    char uuid_str[AWS_UUID_STR_LEN] = {0};
+    struct aws_byte_buf uuid_buf = aws_byte_buf_from_array(uuid_str, sizeof(uuid_str));
+    uuid_buf.len = 0;
+    aws_uuid_to_str(&uuid, &uuid_buf);
+
+    memcpy(result, AWS_LITERAL_APPLE_DISPATCH_QUEUE_ID_PREFIX, AWS_IO_APPLE_DISPATCH_QUEUE_ID_PREFIX_LENGTH);
+    memcpy(result + AWS_IO_APPLE_DISPATCH_QUEUE_ID_PREFIX_LENGTH, uuid_buf.buffer, uuid_buf.len);
+}
+
+/* Setup a dispatch_queue with a scheduler. */
+struct aws_event_loop *aws_event_loop_new_with_dispatch_queue(
+    struct aws_allocator *alloc,
+    const struct aws_event_loop_options *options) {
+    AWS_PRECONDITION(options);
+    AWS_PRECONDITION(options->clock);
+
+    struct aws_event_loop *loop = aws_mem_calloc(alloc, 1, sizeof(struct aws_event_loop));
+    struct dispatch_loop *dispatch_loop = NULL;
+
+    AWS_LOGF_DEBUG(AWS_LS_IO_EVENT_LOOP, "id=%p: Initializing dispatch_queue event-loop", (void *)loop);
+    if (aws_event_loop_init_base(loop, alloc, options->clock)) {
+        goto clean_up;
+    }
+
+    dispatch_loop = aws_mem_calloc(alloc, 1, sizeof(struct dispatch_loop));
+    dispatch_loop->allocator = alloc;
+    loop->impl_data = dispatch_loop;
+    dispatch_loop->base_loop = loop;
+
+    char dispatch_queue_id[AWS_IO_APPLE_DISPATCH_QUEUE_ID_LENGTH] = {0};
+    s_get_unique_dispatch_queue_id(dispatch_queue_id);
+
+    dispatch_loop->dispatch_queue = dispatch_queue_create(dispatch_queue_id, DISPATCH_QUEUE_SERIAL);
+    if (!dispatch_loop->dispatch_queue) {
+        AWS_LOGF_FATAL(AWS_LS_IO_EVENT_LOOP, "id=%p: Failed to create dispatch queue.", (void *)loop);
+        aws_raise_error(AWS_ERROR_SYS_CALL_FAILURE);
+        goto clean_up;
+    }
+
+    AWS_LOGF_INFO(
+        AWS_LS_IO_EVENT_LOOP, "id=%p: Apple dispatch queue created with id: %s", (void *)loop, dispatch_queue_id);
+
+    aws_mutex_init(&dispatch_loop->synced_data.lock);
+    dispatch_loop->synced_data.is_executing = false;
+
+    int err = aws_task_scheduler_init(&dispatch_loop->scheduler, alloc);
+    if (err) {
+        AWS_LOGF_ERROR(AWS_LS_IO_EVENT_LOOP, "id=%p: Initializing task scheduler failed", (void *)loop);
+        goto clean_up;
+    }
+
+    aws_linked_list_init(&dispatch_loop->synced_data.cross_thread_tasks);
+
+    struct dispatch_loop_context *context = aws_mem_calloc(alloc, 1, sizeof(struct dispatch_loop_context));
+    aws_ref_count_init(&context->ref_count, context, s_dispatch_loop_context_destroy);
+    context->allocator = alloc;
+    aws_mutex_init(&context->scheduling_state.services_lock);
+
+    if (aws_priority_queue_init_dynamic(
+            &context->scheduling_state.scheduled_services,
+            alloc,
+            DEFAULT_QUEUE_SIZE,
+            sizeof(struct scheduled_service_entry *),
+            &s_compare_timestamps)) {
+        goto clean_up;
```

Comment:
If this fails and we goto clean_up, `dispatch_loop_context *context` has not yet been assigned to dispatc_loop->context. This means when `s_dispatch_event_loop_destroy(loop)` checks whether it should decref `context` by doing a NULL check, it fails and leaks memory. We should assign `dispatch_loop->context` with this `context` immediately following its ref_count being initialized so it gets cleaned properly.